File: lib/entity/event_list.rb
module Synthea
  class EventList

    EMPTY = [].freeze

    attr_accessor :events     # Hash of Synthea::Event

    def initialize
      @events = {}
      @events[:all] = []
      @events[:unprocessed] = []
    end

    def unprocessed
      @events[:unprocessed]
    end

    def process(event)
      event.processed = true
      @events[:unprocessed].delete(event)
    end

    def unprocessed_before(date,type)
      @events[:unprocessed].select{|e|e.type==type && e.time <= date}
    end

    def unprocessed_since(date,type)
      @events[:unprocessed].select{|e|e.type==type && e.time >= date}
    end

    def before(date,type=:all)
      list = @events[type]
      return EMPTY if list.nil? || list.empty?
      end_index = list.index{|e|e.time >= date}
      end_index = list.length if end_index.nil?
      if !end_index.nil?
        list[0..(end_index-1)]
      else
        EMPTY
      end
    end

    def since(date,type=:all)
      list = @events[type]
      return EMPTY if list.nil? || list.empty?
      end_index = list.index{|e|e.time >= date}
      if !end_index.nil?
        list[end_index..list.length]
      else
        EMPTY
      end
    end

    def create(time, type, rule, processed=false)
      @events[type] = [] if @events[type].nil?
      event = Synthea::Event.new(time, type, rule, processed)
      @events[type] << event
      @events[:all] << event
      @events[:unprocessed] << event if !processed
    end

  end
end

File: src/main/java/org/mitre/synthea/engine/EventList.java
package org.mitre.synthea.engine;

import java.util.ArrayList;
import java.util.List;

public class EventList {
  private List<Event> events = new ArrayList<Event>();
  private final Object lock = new Object();

  /**
   * Get the last event of the given type.
   * 
   * @param type
   *          : the type of event
   * @return the last Event of the given type.
   */
  public Event event(String type) {
    Event retVal = null;
    synchronized (lock) {
      for (int i = events.size() - 1; i >= 0; i--) {
        Event event = events.get(i);
        if (event.type.equals(type)) {
          retVal = event;
          break;
        }
      }
    }
    return retVal;
  }

  public void create(long time, String type, String rule, boolean processed) {
    Event event = new Event(time, type, rule, processed);
    synchronized (lock) {
      events.add(event);
    }
  }

  /**
   * Get all events before the given time.
   * 
   * @param time
   *          : the cut off date
   * @return non-null list of events before the cutoff date.
   */
  public List<Event> before(long time) {
    List<Event> retVal = new ArrayList<Event>();
    synchronized (lock) {
      for (Event event : events) {
        if (event.time <= time) {
          retVal.add(event);
        } else if (event.time > time) {
          break;
        }
      }
    }
    return retVal;
  }

  /**
   * Get all events of given type before the given time.
   * 
   * @param time
   *          : the cut off date
   * @param type
   *          : the type of event
   * @return non-null list of events before the cutoff date.
   */
  public List<Event> before(long time, String type) {
    List<Event> retVal = new ArrayList<Event>();
    synchronized (lock) {
      for (Event event : events) {
        if (event.type.equals(type) && event.time <= time) {
          retVal.add(event);
        } else if (event.time > time) {
          break;
        }
      }
    }
    return retVal;
  }

  /**
   * Get all events after the given time.
   * 
   * @param time
   *          : the cut off date
   * @return non-null list of events after the cutoff date.
   */
  public List<Event> after(long time) {
    List<Event> retVal = new ArrayList<Event>();
    synchronized (lock) {
      for (Event event : events) {
        if (event.time >= time) {
          retVal.add(event);
        }
      }
    }
    return retVal;
  }

  /**
   * Get all events of given type after the given time.
   * 
   * @param time
   *          : the cut off date
   * @param type
   *          : the type of event
   * @return non-null list of events after the cutoff date.
   */
  public List<Event> after(long time, String type) {
    List<Event> retVal = new ArrayList<Event>();
    synchronized (lock) {
      for (Event event : events) {
        if (event.time >= time && event.type.equals(type)) {
          retVal.add(event);
        }
      }
    }
    return retVal;
  }

  public String toString() {
    return String.format("EventList (%d events)", events.size());
  }

}
